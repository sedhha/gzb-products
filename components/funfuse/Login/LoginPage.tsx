import React, { useCallback, useEffect, useState } from 'react';
import Auth from '@firebase-client/client.config';
import { signInWithEmailAndPassword } from 'firebase/auth';
import { loginUser } from '@redux-slices/user';
import { useAppDispatch, useAppSelector } from '@redux-tools/hooks';
import Head from 'next/head';
import ThemeLogo from '@/components/funfuse/Logos/ThemeLogo';
import ThemeInputBox from '@/components/common/Inputs/ThemeInputBox';
import { FirebaseError } from 'firebase/app';
import ThemeButton from '../Buttons/ThemeButton/ThemeButton';
import { firebaseErrorTranslater } from '@firebase-server/public/errorTranslator';
import { useRouter } from 'next/router';
import {
  genericRoutes,
  verifiedRoutes,
} from '@/components/funfuse/constants/verifiedRoutes';
import Link from 'next/link';
import useAbortableState from '@hooks/useAbortableState';

type Props = {
  idToken: string;
};
export default function LoginPage({ idToken }: Props) {
  const [email, setEmail] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [error, setError] = useState<string>('');
  const [loading, setLoading] = useState<boolean>(false);
  const [mounted, setMounted] = useState<boolean>(true);

  useEffect(() => {
    return () => {
      setMounted(false);
    };
  }, []);
  const dispatch = useAppDispatch();
  const router = useRouter();
  const { user, isUserVerified } = useAppSelector((state) => state.user);

  const errorHandler = (error: FirebaseError) => {
    console.log('Error while logging in = ', error);
    if (error.code !== undefined && mounted)
      setError(firebaseErrorTranslater(error.code));
    else mounted && setError('Unexpected Error Occured. Try Again');
    if (mounted) setLoading(false);
  };

  const successHandler = useCallback(
    (token: string) => {
      dispatch(loginUser(token));
      mounted && setError('');
      mounted && setLoading(false);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [dispatch, setError, setLoading]
  );

  useEffect(() => {
    if (idToken !== '') {
      setLoading(true);
      successHandler(idToken);
    }
  }, [idToken, successHandler, setLoading]);

  useEffect(() => {
    if (user?.username) {
      if (isUserVerified)
        router.push(`/funfuse/${user.username}/${verifiedRoutes.HOME_ROUTE}`);
      else router.push(`/funfuse/${genericRoutes.UNVERIFIED}`);
    }
  }, [user?.username, router, isUserVerified]);

  const loginHander = () => {
    setLoading(true);
    signInWithEmailAndPassword(Auth, email, password)
      .then((userRecord) => {
        userRecord.user.getIdToken().then(successHandler).catch(errorHandler);
      })
      .catch(errorHandler);
  };

  const isError = error !== '';

  return (
    <React.Fragment>
      <Head>
        <title>Login to Funfuse</title>
      </Head>
      <div className='flex flex-col justify-center w-screen h-screen p-2'>
        <div className='h-[10rem] w-auto'>
          <ThemeLogo />
        </div>
        <ThemeInputBox
          iconClass='funfuse-icons-email'
          placeholder='Enter Email Address'
          onChange={(e) => setEmail(e.target.value)}
          value={email}
          type={'email'}
        />
        <ThemeInputBox
          iconClass='funfuse-icons-lock'
          placeholder='Enter Password'
          onChange={(e) => setPassword(e.target.value)}
          value={password}
          type={'password'}
        />
        {isError ? (
          <label className='m-2 mt-4 text-red-400 font-funfuse'>{error}</label>
        ) : null}
        {loading ? (
          <div className='w-16 h-16 mx-auto mt-8 border-b-2 rounded-full border-funfuse animate-spin' />
        ) : (
          <ThemeButton
            buttonText={'Login'}
            twClass='w-3/4 mx-auto text-xl mt-8'
            buttonCallback={loginHander}
          />
        )}
        <label className='m-2 mt-4 text-gray-400 font-funfuse'>
          Don&apos;t have an Account ?{' '}
          <Link href='/funfuse/register' passHref>
            <a className='underline cursor-pointer text-funfuse'>
              Register Here.
            </a>
          </Link>
        </label>
      </div>
    </React.Fragment>
  );
}
