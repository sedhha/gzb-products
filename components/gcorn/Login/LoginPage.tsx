import React from 'react';
import Head from 'next/head';
import classes from '../Home.module.scss';
import Image from 'next/image';
import InputBox from '../../common/Inputs/InputBox';
import AppConstants from '../global/constants/constants';
import OutlineButton from '@/components/common/Buttons/OutlineButton';
export default function GCornHomePage() {
  return (
    <React.Fragment>
      <Head>
        <title>gCorn: A Corner for Geeks</title>
        <link rel='icon' href='/gcorn/gcorn-logo.svg' />
      </Head>
      <main className={classes.mainSection}>
        <div className={'image-container'}>
          <Image
            src={'/gcorn/gcorn-logo.svg'}
            alt={'gcorn-logo'}
            layout='fill'
            objectFit='contain'
          />
        </div>
        <div className={classes.WrapperDiv}>
          <h1 className={classes.Header}>gCorn!</h1>
          <p className={classes.Tagline}>A Corner for Geeks</p>
        </div>
        <br />
        <InputBox
          placeholder={'Email Address'}
          outlineColor={AppConstants.THEME_GREEN}
        />
        <InputBox
          placeholder={'Password'}
          type='password'
          outlineColor={AppConstants.THEME_GREEN}
        />
        <br />
        <OutlineButton outlineColor={AppConstants.THEME_GREEN}>
          Sign In
        </OutlineButton>
      </main>
    </React.Fragment>
  );
}
