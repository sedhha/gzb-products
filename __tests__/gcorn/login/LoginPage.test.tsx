import LoginPage from '@/components/gcorn/login/LoginPage';
import { render } from '@testing-library/react';
describe('LoginPage Test', () => {
  it('should render without throwing an error', () => {
    const { container } = render(<LoginPage />);
    expect(container).toMatchSnapshot();
  });
});
