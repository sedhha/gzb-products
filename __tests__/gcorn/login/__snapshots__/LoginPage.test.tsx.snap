// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage Test should render without throwing an error 1`] = `
<div>
  <main
    class="mainSection"
  >
    <div
      class="image-container"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="gcorn-logo"
          data-nimg="fill"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
        />
        <noscript />
      </span>
    </div>
    <div
      class="WrapperDiv"
    >
      <h1
        class="Header"
      >
        gCorn!
      </h1>
      <p
        class="Tagline"
      >
        A Corner for Geeks
      </p>
    </div>
    <br />
    <input
      class="sc-bdvvtL sc-gsDKAQ fpvgJU ftImbR"
      placeholder="Email Address"
    />
    <input
      class="sc-bdvvtL sc-gsDKAQ fpvgJU ftImbR"
      placeholder="Password"
      type="password"
    />
    <br />
    <button
      class="sc-dkPtRN sc-hKwDye bNgrzi dlxXti"
    >
      Sign In
    </button>
  </main>
</div>
`;
